Include/internal/pycore_code.h

```diff
@@ -30,6 +30,10 @@ typedef struct {
     uint32_t builtin_keys_version;
 } _PyLoadGlobalCache;
 
+typedef struct {
+    PyObject *obj;  /* borrowed */
```

Comment:
Is this safe? 
I think it is, as long as all reads are guarded by type version checks, *and* the obj belongs to the class (or a super-class) being versioned checked, *and* they is no way for the state of the class (or any of its superclasses) to change.

Could you add a (reasonably detailed) explanation of why this is safe, and a clear warning of what pre-conditions are necessary to keep it safe.

Reply:
Yes. I thought quite hard about this before going ahead 😉 .

In the case of methods, there will always be a reference to ``obj`` held by the type object, so it will never be cleared by GC. Should the method be deleted, e.g. ``del X.y``, then the type cache will be invalidated and its version tag updated. We only need to check the type version tag.

The type method cache uses the exact same principles -- it's a 4096(?) element array of borrowed method objects! We can make use of all the heavy lifting it's already done for us in this area -- ``tp_version_tag`` is always invalidated on MRO modification, and on every ``type_set_xx`` operation (see ``PyType_Modified`` usages). So if it's somehow lasted from 2.6 till present, I'd think it's battle-tested enough.

Reply:
Added at the specialization site instead of the struct declaration:
https://github.com/python/cpython/pull/27722/commits/afcf51e67496ce86d5c045d329d8d0a7832885c2#diff-22e653d878778ca28317261f7826545288d02fd9980cf485d0329f28a34993f2R871